### ISSUE TITLE
Debug Failure. Parameter symbol already has a cached type which differs from newly assigned type.

### ISSUE THREAD

**Issue by albertodvc** [microsoft/TypeScript#50773]
# Bug Report

### 🔎 Search Terms

Error: Debug Failure. Expected [object Object] === [object Object]. Parameter symbol already has a cached type which differs from newly assigned type

### 🕗 Version & Regression Information

- This is a crash
- This changed between versions 4.6.4 and 4.7.2
- This is still happening with the currently latest 4.8.3

### ⏯ Playground Link

Can't reproduce in the playgorund. To reproduce, run:
```sh
tsc code.js --allowJs --noEmit --target esnext.
```

### 💻 Code

code.js

```js
class Model {
  save() {
    return this.createMissingLabels()
      .then(() => this.$store.save(this));
  }

  createMissingLabels() {
    if (this.labels && Array.isArray(this.labels)) {
      return Promise.all(this.labels.map(
        label => this.$store.createLabel(label)
      )).then(labels => (this.labels = labels));
    }
    return Promise.resolve();
  }
}

```

### 🙁 Actual behavior

```sh
Debug Failure. Expected [object Object] === [object Object]. Parameter symbol already has a cached type which differs from newly assigned type
```

### 🙂 Expected behavior

No crash

**Comment by jakebailey** [comment]
Unsurprisingly, this bisects to #43845, so this appears to be another case where we are accidentally fixing a parameter more than once, but what must be a different code path than #49740.

**Comment by smleake** [comment]
Seeing this same issue on 4.7.2. Switching to 4.6.4 remedied the problem, but definitely less than ideal.

**Comment by jhmaster2000** [comment]
I've just ran into this issue on `4.9.3` and while it for some reason fails to reproduce on TSPlayground I did manage to make it reproduce on the TS Bug Workbench. Interestingly, it does not reproduce on the current **Nighty** (`5.0.0-dev.20221130`) and instead produces proper TS errors:

- [4.9.3 Bug Workbench](https://www.typescriptlang.org/dev/bug-workbench/?ts=4.9.3#code/LAKA5gNg9gRghhAKgCwJYGcB0B5GArAUwGMAXAAgF4yA7AgdzIAUAnKADwE8AKXQ0gGjIBvUGTFkwBElxJxmkkgHoAVAC4ycah2WLBAB1Z71AawIcoAMzIkOegpetyFASmGjxHxYrIA9APzuHmJeZACSViTIBGQEbHoQqESo5Joc1rbRGGQArtREUAC2BQTUJAQAJtZRMcyszBJS6GQWqGwVmB2BQcxS2czUjvJSANoGUHoAugDcXQC+-KCzzjMgQA)
- [Nightly Bug Workbench](https://www.typescriptlang.org/dev/bug-workbench/?ts=5.0.0-dev.20221130#code/LAKA5gNg9gRghhAKgCwJYGcB0B5GArAUwGMAXAAgF4yA7AgdzIAUAnKADwE8AKXQ0gGjIBvUGTFkwBElxJxmkkgHoAVAC4ycah2WLBAB1Z71AawIcoAMzIkOegpetyFASmGjxHxYrIA9APzuHmJeZACSViTIBGQEbHoQqESo5Joc1rbRGGQArtREUAC2BQTUJAQAJtZRMcyszBJS6GQWqGwVmB2BQcxS2czUjvJSANoGUHoAugDcXQC+-KCzzjMgQA)

**Comment by danthegoodman** [comment]
I ran into this when upgrading TS from 4.2.3 to 4.9.3 on a large internal project. I can't boil it down to a reproducable set of code, but here are the steps I followed to find the offending section of code so I could refactor how the types were specified.

1. Run `tsc` with `--generateTrace traceOutputDir`
2. Then run `npx @typescript/analyze-trace traceOutputDir`

At the top of the output was a path to the file that was last processed. I was then able to comment out all of the code in the file, verify that the compile was successful and then slowly reintroduce code until necessary.

In my case, I was able to add an explicit type to a generic function that was trying to do type inference.

You can read more about that analyze-trace tool [here](https://github.com/microsoft/TypeScript-wiki/blob/main/Performance-Tracing.md)

**Comment by marekdedic** [comment]
Hi,
I'm running into this issue here: https://github.com/skaut/wordpress-version-checker/actions/runs/6865709436/job/18670310204?pr=1051

I know the file where the issue happens ([here](https://github.com/skaut/wordpress-version-checker/blob/a2160e483d3ed5af90ba032a85ab58eb14c2a300/__tests__/setup.ts)), but have no idea how to fix this. And unlike others, I'm on TS 5.2.2

Happy to help investigating this, just don't really know how. On my local machine, it's a heisenbug :(

**Comment by andrewbranch** [comment]
This was fixed in TypeScript 5.0. @marekdedic if you can still reproduce a bug with the same error message, please file a new issue with some repro steps. The logs from the CI run you linked have expired. Thanks!

**Comment by marekdedic** [comment]
Hi @andrewbranch 

Thanks for the response, unfortunately, I am still getting this with TS 5.3. I filed a new issue (#57435) with a repro. Could you please take a look at it there and let me know if there's anything more you need?
